core/schema/directory.go

```diff
@@ -343,26 +346,98 @@ type dirDockerBuildArgs struct {
 	Secrets    []core.SecretID                    `default:"[]"`
 }
 
-func (s *directorySchema) dockerBuild(ctx context.Context, parent *core.Directory, args dirDockerBuildArgs) (*core.Container, error) {
-	platform := parent.Query.Platform()
+func getDockerIgnoreFileContent(ctx context.Context, parent dagql.Instance[*core.Directory], filename string) ([]byte, error) {
+	file, err := parent.Self.File(ctx, filename)
+	if err != nil {
+		if strings.Contains(err.Error(), "no such file or directory") {
+			return nil, nil
+		}
+
+		return nil, err
+	}
+
+	content, err := file.Contents(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return content, nil
+}
+
+func (s *directorySchema) dockerBuild(ctx context.Context, parent dagql.Instance[*core.Directory], args dirDockerBuildArgs) (*core.Container, error) {
+	platform := parent.Self.Query.Platform()
 	if args.Platform.Valid {
 		platform = args.Platform.Value
 	}
-	ctr, err := core.NewContainer(parent.Query, platform)
+
+	var dockerIgnoreContents []byte
+	var err error
+
+	// use dockerfile specific .dockerfile if that exists
+	if args.Dockerfile != "Dockerfile" {
+		// https://docs.docker.com/build/concepts/context/#filename-and-location
+		specificDockerIgnoreFile := args.Dockerfile + ".dockerignore"
+		dockerIgnoreContents, err = getDockerIgnoreFileContent(ctx, parent, specificDockerIgnoreFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// fallback on default .dockerignore file
+	if len(dockerIgnoreContents) == 0 {
+		dockerIgnoreContents, err = getDockerIgnoreFileContent(ctx, parent, ".dockerignore")
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	excludes, err := ignorefile.ReadAll(bytes.NewBuffer(dockerIgnoreContents))
 	if err != nil {
 		return nil, err
 	}
+
+	var destdir dagql.Instance[*core.Directory]
+	if len(excludes) > 0 {
+		err = s.srv.Select(ctx, parent, &destdir,
+			dagql.Selector{
+				Field: "withDirectory",
+				Args: []dagql.NamedInput{
+					{Name: "path", Value: dagql.NewString("./filtered")},
```

Comment:
This can just be the empty string, which puts it at the root, so we don't need the second selector :tada:

Reply:
hmm i did tried this but that returns the dir with the excluded files (I think because src and dest dir are same 🤔 from llb perspective?)

i renamed the dir from `filtered` to `buildctx` and also added a comment to make it clearer why we are doing this.

Reply:
Hi @jedevc, i think this is the only blocking comment from merging this code. please let me know your thoughts about this one. 🙏

Reply:
Hi @jedevc, a kind reminder for this comment.

Reply:
I think I can use the new `dir.filter` api (https://github.com/dagger/dagger/pull/9976) to simplify this further. but do we want to hold off until that api is merged?

Reply:
Just merged it, feel free to use it :tada: